### PR TITLE
CATROID-568 REFACTOR ChangeVariableActionTest to Kotlin

### PR DESCRIPTION
**Pull request by mandelbrot2357**
Converted test to Kotlin, added parameterization

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            arrayOf(Formula(CHANGE_VARIABLE_VALUE), null, INITIALIZED_VALUE),
+            arrayOf(Formula(CHANGE_VARIABLE_VALUE.toString()), UserVariable(TEST_USERVARIABLE, INITIALIZED_VALUE), EXPECTED_NEW_VALUE),
+            arrayOf(Formula(NOT_NUMERICAL_STRING), UserVariable(TEST_USERVARIABLE, INITIALIZED_VALUE), INITIALIZED_VALUE),
+            arrayOf(null, UserVariable(TEST_USERVARIABLE, INITIALIZED_VALUE), INITIALIZED_VALUE),
+            arrayOf(Formula(Double.NaN), UserVariable(TEST_USERVARIABLE, INITIALIZED_VALUE), INITIALIZED_VALUE)
+        )
+
+        private const val NOT_NUMERICAL_STRING = "changeVariable"
+        private const val TEST_USERVARIABLE = "testUservariable"
+        private val CHANGE_VARIABLE_VALUE = ThreadLocalRandom.current().nextDouble(-100.0, 100.0)
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt:59]: Please don't use random values for testing.
Instead, run a whole range like this as parameters: 
IntRange(0, 99).flatMap { ... } (don't forget the .toDouble())

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        private val INITIALIZED_VALUE = ThreadLocalRandom.current().nextDouble(-100.0, 100.0)
+        private val EXPECTED_NEW_VALUE = INITIALIZED_VALUE + CHANGE_VARIABLE_VALUE
+    }
+
+    @Before
+    fun setUp() {
+        testSprite = SingleSprite("testSprite")
+        project = Project(MockUtil.mockContextForProject(), "testProject")
+        ProjectManager.getInstance().currentProject = project
+        userVariable = variable ?: UserVariable(TEST_USERVARIABLE, INITIALIZED_VALUE)
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt:69]: The "userVariable" is always UserVariable(TEST_USERVARIABLE, INITIALIZED_VALUE) except test-case 1.
As discussed, please edit the test-case and delete this variable.

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+@RunWith(Parameterized::class)
+class ChangeVariableActionTest(private val formula: Formula?, private val variable: UserVariable?, private val expectedValue: Double) {
+    private lateinit var testSprite: Sprite
+    private lateinit var userVariable: UserVariable
+    private lateinit var project: Project
+
+    companion object {
+        @JvmStatic
+        @Parameterized.Parameters
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt:47]: Please add a name for the test-cases.

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.java [hunk 4]
```diff
@@ -1,94 +0,0 @@
...
-import org.catrobat.catroid.test.MockUtil;
-import org.junit.Before;
-import org.junit.Test;
-import org.junit.runner.RunWith;
-import org.junit.runners.JUnit4;
-
-import static junit.framework.Assert.assertEquals;
-
-@RunWith(JUnit4.class)
-public class ChangeVariableActionTest {
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.java:40]: Great! It looks much better now!

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        testSprite = SingleSprite("testSprite")
+        project = Project(MockUtil.mockContextForProject(), "testProject")
+        ProjectManager.getInstance().currentProject = project
+        userVariable = variable ?: UserVariable(TEST_USERVARIABLE, INITIALIZED_VALUE)
+        project.addUserVariable(userVariable)
+    }
+
+    @Test
+    fun testChangeUserVariableAndCheckResult() {
+        testSprite.actionFactory.createChangeVariableAction(testSprite, formula, variable).act(1f)
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt:75]: Please also verify that "act" always returns true.

Comment (gPathpp) [reply catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt:75]: Maybe in an extra test-method, if you like.

**Comment by mandelbrot2357** [comment]
The current test doesn't contain any randomization. Did you maybe look at an older commit? The idea behind multiple values was to catch bugs that are value-dependent, but I could use reduce it to just -1, 0, 1 if runtime is an issue. The parameterization is meant to get rid of the previous copy-pasted code.

**Comment by wslany** [comment]
Regarding randomization, you are right that there's none, I was just mislead by the first commit message above 
image

**Comment by mandelbrot2357** [comment]
Oops. Fixed it

**Comment by wslany** [comment]
It's not so much about the runtime, though runtime also accumulates over the large number of tests, but I just want to have value for each test. The test should also serve as a documentation of the function. I almost think that these tests should all be replaced by Catrobat language tests only.

**Comment by wslany** [comment]
Catrobat language tests have the advantage that they are easier to read (hopefully, and soonish), and also help other teams, e.g., the Catty team.

**Comment by mandelbrot2357** [comment]
That's a good point. Do you want to convert it to a Catrobat test?

**Comment by mandelbrot2357** [comment]
I could also just replace the flatMap with fixed values for now and convert it when the parameterization brick is ready.

**Comment by wslany** [comment]
> I could also just replace the flatMap with fixed values for now and convert it when the parameterization brick is ready.

Sounds good. Most likely we'll rewrite a lot of them to Catrobat language tests (CLTs) once the parameterization brick is ready, but until it is OK to have it rewritten to Kotlin. So yes, please proceed.

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+) {
+    private lateinit var testSprite: Sprite
+
+    companion object {
+        @JvmStatic
+        @Parameterized.Parameters(name = "{0}")
+        fun parameters() = listOf(
+            arrayOf("INVALID_USER_VARIABLE", Formula(CHANGE_VARIABLE_VALUE), null, null),
+            arrayOf("NUMBER_NEGATIVE", Formula(CHANGE_VARIABLE_VALUE), UserVariable(TEST_USERVARIABLE, -1.0), -1.0 + CHANGE_VARIABLE_VALUE),
+            arrayOf("NUMBER_ZERO", Formula(CHANGE_VARIABLE_VALUE), UserVariable(TEST_USERVARIABLE), + CHANGE_VARIABLE_VALUE),
```
Comment (wslany) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/ChangeVariableActionTest.kt:53]:
```suggestion
            arrayOf("NUMBER_ZERO", Formula(CHANGE_VARIABLE_VALUE), UserVariable(TEST_USERVARIABLE), CHANGE_VARIABLE_VALUE),
```

**Comment by wslany** [comment]
Thanks!

**Pull request merged**